# Calculating composite MCWD and MAP values

In Duque et al. (2019), we have created a new kind of plot in which we display together the maximum climatological water deficit (MCWD) and mean annual precipitation (MAP) differences between two experiment simulations. We believe this kind of plot shows us valuable information about what land world regions are more vulnerable to experiment scenarios. In our experimental module we have a function to help us create this plot. Let's import obrero and the module:

In [1]:
# small hack to be able to import module without install
import os
import sys
sys.path.append(os.getcwd() + '/../')

import obrero
from obrero.experimental import mcwd

Now we need data from two simulations. As well as an evaporation file. We always use the evaporation values from the control simulation because we think those values should not change if land cover stays the same. That is to say that energy demand from what is an ecosystem should not change if the ecosystem is to be preserved:

In [3]:
# file name
f1 = 'data/ctl_pr_evap.nc'
f2 = 'data/pen_pr_evap.nc'
f3 = 'data/ctl_evapmean.nc'

# read as data array (ignore warnings because netCDF files are weird)
da1 = obrero.read_nc(f1, 'pr')
da2 = obrero.read_nc(f2, 'pr')
ev = obrero.read_nc(f3, 'evap')

First we have to get MAP, which is the sum of all precipitation in every year:

In [4]:
map1 = da1.groupby('time.year').sum(dim='time', keep_attrs=True)
map2 = da2.groupby('time.year').sum(dim='time', keep_attrs=True)

# rename time
map1 = map1.rename({'year': 'time'})
map2 = map2.rename({'year': 'time'})

Next we calculate MCWD for each experiment using the same evaporation in both of them:

In [5]:
# compute mcwd
wd1 = mcwd.get_mcwd(da1, ev)
wd2 = mcwd.get_mcwd(da2, ev)

To get the composite map values we use function `mcwd_composite_map()` in the experimental `obrero.experimental.mcwd` module. The order of the arrays matters: first MCWD and MAP from an experiment and then the same for the **control** experiment. This is because there will be substractions. It will return a gridded map as well as a data frame with all values tabulated:

In [9]:
composite, table = mcwd.mcwd_composite_map(wd2, map2, wd1, map1)

Let's look at these two things:

In [10]:
composite

<xarray.DataArray 'composite' (latitude: 32, longitude: 64)>
array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])
Coordinates:
  * latitude   (latitude) float64 85.76 80.27 74.74 ... -74.74 -80.27 -85.76
  * longitude  (longitude) float64 0.0 5.625 11.25 16.88 ... 343.1 348.8 354.4
Attributes:
    long_name:  Composite MAP and MCWD
    units:      1

This is the map which is used for plotting purposes. And then the table:

In [11]:
table

lon        lat      ctl_map      exp_map    ctl_mcwd    exp_mcwd  comp
id                                                                            
0   236.250  69.212976   568.175017   628.333967  -89.667515  -67.413326   5.5
1   241.875  69.212976   584.077378   644.068658  -76.084100  -59.616325   5.5
2    39.375  58.142954   791.212306   886.235404  -68.037802  -29.971604   5.5
3    22.500  52.606526   790.350410   891.052641 -184.560142 -119.409784   5.5
4    33.750  52.606526   419.365649   484.679045 -158.039537 -134.594018   5.5
5    39.375  52.606526   522.831663   592.652493 -181.348686 -149.360635   5.5
6    50.625  52.606526   585.916357   666.182773 -122.353594 -102.642326   5.5
7    95.625  52.606526  1087.879558  1216.880009  -12.081088   -1.436797   5.5
8     5.625  47.069642   684.800756   807.632279 -135.093102  -71.060846   5.5
9    11.250  47.069642  1080.181526  1214.903778 -138.299875  -48.440723   5.5
10  135.000  47.069642  1006.774190  1135.454262 -186.813532 -133.437255   5.5
11  236.250  47.069642  1258.221005  1391.932856 -194.132246 -132.932938   5.5
12  241.875  47.069642   946.491523  1121.291573 -217.914480 -133.803777   5.5
13  253.125  35.995078   667.306753   828.508314 -182.683743 -107.942736   5.5
14   56.250  30.457554    25.491783    64.624925  -15.010287   -8.270842   5.5
15  247.500  30.457554    76.621196   104.876924  -40.470625  -25.413133   5.5
16  253.125  30.457554   341.589670   509.780974 -141.349765  -74.214104   5.5
17  258.750  30.457554   850.434282  1008.257597 -112.265136  -55.182284   5.5
18  270.000  30.457554   297.528154   393.638259  -78.319618  -36.603156   5.5
19   56.250  24.919929    39.417019    61.195551  -18.543463  -11.150562   5.5
20  253.125  24.919929   375.931582   597.789419 -165.786545 -105.695605   5.5
21   95.625  19.382231  1006.844215   844.645020 -531.893910 -587.550197   0.5
22  270.000  19.382231  1163.173692   859.185751 -318.991009 -461.152327   0.5
23  275.625  13.844484  2160.137078  1362.844290 -356.598691 -785.470956   0.5
24  286.875   8.306703  1438.812212  1158.435547 -396.477508 -530.568300   0.5
25   16.875   2.768903  1702.625313  1914.317191 -430.654809 -353.120016   5.5
26   22.500   2.768903  1471.878639  1648.201893 -490.944782 -393.826922   5.5
27  101.250   2.768903  1663.691018  1874.155965 -301.413738 -198.759283   5.5
28  292.500  -2.768903  1820.507271  1493.382756 -297.218984 -464.353205   0.5
29  303.750  -2.768903  2859.418053  2397.780667  -63.090608 -128.911392   0.5
30  309.375  -2.768903  1825.769391  1547.958302 -293.414998 -461.674991   0.5
31  315.000  -2.768903  2114.041411  1869.875377 -329.453544 -505.344410   0.5
32  140.625  -8.306703  3195.035533  2166.047030  -21.770042 -144.420468   0.5
33  298.125  -8.306703  2255.774588  2081.171569 -406.579731 -498.399087   2.5
34  303.750  -8.306703  1912.524569  1659.916127 -453.611176 -524.544758   0.5
35  315.000  -8.306703  1009.021342   852.484924 -263.267415 -355.428017   0.5
36  320.625  -8.306703  1428.933074  1019.331659 -203.423587 -454.735884   0.5
37  129.375 -13.844484   991.249235   591.708556 -462.478020 -564.441713   0.5
38  135.000 -13.844484  1064.762543   522.172483 -437.192214 -631.485210   0.5
39  309.375 -13.844484  1250.270434  1094.161356 -449.990502 -522.360920   0.5
40  315.000 -13.844484  1537.537901  1377.286635 -454.121243 -608.828766   0.5
41  123.750 -19.382231   774.145905   447.911488 -344.538737 -449.713240   0.5
42  135.000 -19.382231   799.054668   508.657273 -292.043783 -362.989423   0.5
43  140.625 -19.382231   653.610432   255.850463 -239.799580 -365.057603   0.5
44  146.250 -19.382231   771.725022   230.817059 -245.277103 -433.868728   0.5
45  118.125 -24.919929   406.300153   202.703221 -215.000199 -269.246937   0.5
46  123.750 -24.919929   904.816500   675.028759 -396.660816 -455.576942   0.5
47  129.375 -24.919929   703.153499   447.781300 -241.944332 -326.255394   0.5
48  135.000 -24.919929   357.422872   199.418752 -112.2654

The composite value can be one of 0.5, 1.5, 2.5, 3.5, 4.5, or 5.5. The meaning of each of these values is explained in detail in Duque et al. (2019).

## References

Duque et al. (2019).